dipy/io/statefull_tractogram.py

```diff
@@ -0,0 +1,879 @@
+    """ Enum to simplify future change to convention """
+    VOX = 'vox'
+    VOXMM = 'voxmm'
+    RASMM = 'rasmm'
+
+
+class StateFullTractogram(object):
+    """ Object designed to be identical no matter the file format
+    (trk, tck, fib, dpy). Facilitate transformation between space and
+    data manipulation for each streamline / point.
+    """
+
+    def __init__(self, streamlines, reference, space,
+                 shifted_origin=False,
+                 data_per_point=None, data_per_streamline=None):
+        """ Create a strict, state-aware, robust tractogram
+
+        Parameters
+        ----------
+        streamlines : list or ArraySequence
+            Streamlines of the tractogram
+        reference : Nifti or Trk filename, Nifti1Image or TrkFile,
+            Nifti1Header or trk.header (dict)
+            Reference that provides the spatial attribute.
+            Typically a nifti-related object from the native diffusion used for
+            streamlines generation
+        space : string
+            Current space in which the streamlines are (vox, voxmm or rasmm)
+            Typically after tracking the space is VOX, after nibabel loading
+            the space is RASMM
+        shifted_origin : bool
+            Information on the position of the origin,
+            False is Trackvis standard, default (corner of the voxel)
+            True is NIFTI standard (center of the voxel)
+        data_per_point : dict
+            Dictionary in which each key has X items, each items has Y_i items
+            X being the number of streamlines
+            Y_i being the number of points on streamlines #i
+        data_per_streamline : dict
+            Dictionary in which each key has X items
+            X being the number of streamlines
+
+        Notes
+        -----
+        Very important to respect the convention, verify that streamlines
+        match the reference and are effectively in the right space.
+
+        Any change to the number of streamlines, data_per_points or
+        data_per_streamlines requires particular verification.
+
+        In a case of manipulation not allowed by this object, use Nibabel
+        directly and be careful.
+        """
+        if data_per_point is None:
+            data_per_point = {}
+
+        if data_per_streamline is None:
+            data_per_streamline = {}
+
+        if not isinstance(streamlines, (list, ArraySequence)):
+            raise TypeError('Streamlines MUST be a list or an ArraySequence')
+        self._streamlines = ArraySequence(streamlines)
+
+        if not _is_data_per_point_valid(streamlines, data_per_point):
+            raise ValueError('Invalid data per point, does not match')
+        if not _is_data_per_streamline_valid(streamlines, data_per_streamline):
+            raise ValueError('Invalid data per streamlines, does not match')
+
+        self._data_per_point = dict(data_per_point)
+        self._data_per_streamline = dict(data_per_streamline)
```

Comment:
I'm wondering why aren't you using a `Tractogram` object internally? It would manage the `streamlines`, `data_per_point` and `data_per_streamline` for you and handle the space transformations given that you provide the correct affine in the constructor (i.e. which brings the streamlines in "RAS+ MM" space).

Reply:
I was wondering the same thing, Why your class does not inherit from `Tractogram`?

Reply:
I'll let @frheault answer this, but a quick answer is that using a Tractogram object internally would allow people to access methods such as `Tractogram.affine_to_rasmm()` and `Tractogram.to_world()`, which would leave the `StatefulTractogram` in an inconsistent state.

I guess that's a limitation of Python, since we can't really make anything 100% private.

Inheriting might be a solution, however.

Reply:
I could use tractogram to help with the consistency of streamlines and its data_per_streamlines and data_per_points. However for the spatial handling I find the attribute and function related to tractogram to be potentially confusing if access (what JC said)

After it get in Dipy 1.0 (and some real-life testing), if this get moved to Nibabel. It would be interesting to inherit from Tractogram and potentially LazyTractogram. That redesign/refactor is probably a good idea, but maybe later ?

But for now I will look into Tractogram for data/streamlines handling and keep the internal consistency and spatial transformation handling for the stateful_tractogram